agent/hcp/telemetry_config_provider.go

```diff
@@ -0,0 +1,182 @@
+		return nil, fmt.Errorf("missing HCP client")
+	}
+
+	if opts.MetricsConfig == nil {
+		return nil, fmt.Errorf("missing metrics config")
+	}
+
+	if opts.RefreshInterval <= 0 {
+		return nil, fmt.Errorf("invalid refresh interval")
+	}
+
+	cfg := &dynamicConfig{
+		endpoint:        opts.MetricsConfig.Endpoint,
+		labels:          opts.MetricsConfig.Labels,
+		filters:         opts.MetricsConfig.Filters,
+		refreshInterval: opts.RefreshInterval,
+	}
+
+	hash, err := calculateHash(cfg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to calculate hash: %w", err)
+	}
+
+	t := &telemetryConfigProvider{
+		cfg:       cfg,
+		cfgHash:   hash,
+		logger:    hclog.FromContext(opts.Ctx).Named("telemetry_config_provider"),
+		hcpClient: opts.HCPClient,
+	}
+
+	go t.run(opts.Ctx, opts.RefreshInterval)
+
+	return t, nil
+}
+
+// run continously checks for updates to the telemetry configuration by making a request to HCP.
+// Modification of config only occurs if changes are detected to decrease write locks that block read locks.
+func (t *telemetryConfigProvider) run(ctx context.Context, refreshInterval time.Duration) {
+	ticker := time.NewTicker(refreshInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			if newCfg, hasChanged := t.checkUpdate(ctx); hasChanged {
+				t.modifyTelemetryConfig(newCfg)
+				ticker.Reset(newCfg.refreshInterval)
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// checkUpdate makes a HTTP request to HCP to return a new metrics configuration and true, if config changed.
+// checkUpdate does not update the metricsConfig field to prevent acquiring the write lock unnecessarily.
+func (t *telemetryConfigProvider) checkUpdate(ctx context.Context) (*dynamicConfig, bool) {
+	t.rw.RLock()
+	defer t.rw.RUnlock()
+
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+
+	telemetryCfg, err := t.hcpClient.FetchTelemetryConfig(ctx)
+	if err != nil {
+		t.logger.Error("failed to fetch telemetry config from HCP", "error", err)
+		goMetrics.IncrCounter(internalMetricRefreshFailure, 1)
+		return nil, false
+	}
+
+	newDynamicConfig := &dynamicConfig{
+		filters:         telemetryCfg.MetricsConfig.Filters,
+		endpoint:        telemetryCfg.MetricsConfig.Endpoint,
+		labels:          telemetryCfg.MetricsConfig.Labels,
+		refreshInterval: telemetryCfg.RefreshConfig.RefreshInterval,
+	}
+
+	newHash, err := calculateHash(newDynamicConfig)
+	if err != nil {
+		t.logger.Error("failed to calculate hash for new config", "error", err)
+		goMetrics.IncrCounter(internalMetricRefreshFailure, 1)
+		return nil, false
+	}
+
+	return newDynamicConfig, newHash == t.cfgHash
+}
+
+// modifynewTelemetryConfig acquires a write lock to modify it with a given newTelemetryConfig object.
+func (t *telemetryConfigProvider) modifyTelemetryConfig(newCfg *dynamicConfig) {
+	t.rw.Lock()
+	defer t.rw.Unlock()
+
+	t.cfg = newCfg
+}
+
+// GetEndpoint acquires a read lock to return endpoint configuration for consumers.
```

Comment:
All the Get* methods acquire a reader lock and defer the unlock before returning the config field